### DISCUSSION TITLE
How to validate a list of strings in arbitrary order into a model with ordered attributes?

### DISCUSSION THREAD

**Discussion by soerenlorenz** [pydantic/pydantic#2666] [Question] [answered]
I have these code snippets:
```python
from pydantic import BaseModel

class AttributeOrder(BaseModel):
    pos_x: constr(regex=r'X|x|Position X')
    index_pos_x: int
    pos_y: constr(regex=r'Y|y|Position Y')
    index_pos_y: int
    pos_z: constr(regex=r'Z|z|Position Z')
    index_pos_z: int

# should validate to index_pos values 0, 1, 2
input_1 = [(0, 'x'), (1, 'y'), (2, 'z')]

# should validate to index_pos values 2, 0, 1
input_2 = [(0, 'Y'), (1, 'z'), (2, 'Position X')]

# should validate to index_pos values 1, 3, 0
input_3 = [(0, 'z'), (1, 'x'), (2, 'unecessary'), (3, 'y')]

# should not validate at all
input_4 = [(0, 'z'), (1, 'x'), (2, 'unecessary')]
```
Now I need some way to present the tuples to the class constructor in a matching order or maybe some logic in the ModelClass. One can argue that finding a mapping between input strings and class attributes is out of scope for validation and more a thing of input manipulation. Still it would be nice to have a one-stop solution.

If this is not possible it would help to get the regex strings out of the model definition to not have to write them a second time somewhere in a pre-processing function. I tried this but didn't find them.

I have many different AttributeOrder-like classes and need a generic function to handle the mapping. Simply trying every possible mapping using itertools.permutation on the input does work in principle, but gets slow with raising number of attributes very quickly.

Any ideas?

**Comment by eddsalkield** [comment, marked as answer]
Not quite sure why you want to use a regex to give multiple names to each field (being precise is usually a good thing!), but here's a generic way of unpacking model parameters from a list of tuple:

```python3
from pydantic import BaseModel, root_validator
from typing import List, Tuple, Optional

class UnpackRawModel(BaseModel):
    _to_unpack: List[str] = []
    _raw_data: List[Tuple[int, str]]

    @root_validator(pre=True)
    def unpack_raw_data(cls, values):
        to_unpack = cls.__dict__['_to_unpack']
        for v, k in values['_raw_data']:
            if k in ['_raw_data', '_to_unpack']:
                raise ValueError(f'{k} not a valid field of the model')
            if k not in to_unpack:
                continue
            values[k] = v

        for r in to_unpack:
            if r not in values.keys():
                raise ValueError(f'field {r} required')

        return values

class Coordinate(UnpackRawModel):
    _to_unpack: List[str] = ['x', 'y', 'z']
    x: Optional[int]
    y: Optional[int]
    z: Optional[int]
```

Results:

```python3
>>> Coordinate(_raw_data=[(0, 'x'), (1, 'y'), (2, 'z')])
Coordinate(x=0, y=1, z=2)
>>> Coordinate(_raw_data=[(0, 'z'), (1, 'x'), (2, 'unnecessary'), (3, 'y')])
Coordinate(x=1, y=3, z=0)
>>> Coordinate(_raw_data=[(0, 'z'), (1, 'x'), (2, 'unnecessary')])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/site-packages/pydantic/main.py", line 346, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for Coordinate
__root__
  field y required (type=value_error)
```

By inheriting from `UnpackRawModel`, this behaviour can be used to override any arbitrary parameters - just set them in `_to_unpack`.

**Reply by soerenlorenz** [reply]
Wow. Brilliant. Thank you so much for this solution. And you're right, it's better to not use regexes.

The only thing I definately need is strings in my raw_data which can't be python identifiers. So my model definition could look like:
```python
class CoordinatesHeader(UnpackRawModel):
    _to_unpack: List[Tuple[str, str]] = [('x', 'x'), ('y', 'Pos. Y'), ('z', 'position z')]
    x: int
    y: int
    z: int
```
A little more background: What I'm validating are the column headers of some human created tabular data. I will then use this HeaderModel to load the data of the table rows into a second Pydantic model which will valdiate the actual values. For this I need a mapping of the correct column indexes to the corresponding fields.

I think I can modify your code in this direction. Using root_validator to do some pre-processing before actual validation is the key idea for my question, so I'll mark it answered. Still I would be glad to get any further advice.

**Reply by eddsalkield** [reply]
Glad to be of help!

Extending `_to_unpack` to become a mapping of input identifier to Python identifier makes sense; consider using a dict instead, e.g.:

```python3
_to_unpack: Dict[str, str] = {'Pos. Y': 'y', 'position z': 'z'}
```

The identity transformation (e.g. 'x' -> 'x') can be left implicit.  I'll leave the relevant modifications to utilise this data structure up to you.

Not that I have much context, but if the tabular data you're loading is a csv, consider using a [`csv.DictReader`](https://docs.python.org/3/library/csv.html#csv.DictReader).  It has a nice pattern for extracting information from tables, outputting a `List[Dict]` with the column headers as the keys.
